stdlib/public/runtime/MetadataLookup.cpp

```diff
@@ -755,3 +798,55 @@ swift_getTypeByMangledName(const char *typeNameStart, size_t typeNameLength,
 
   return Demangle::decodeMangledType(builder, node);
 }
+
+void swift::swift_getFieldAt(
+    const Metadata *base, unsigned index,
+    llvm::function_ref<void(std::string &name, FieldType fieldInfo)> callback) {
+  auto &fields = TypeFields.get();
+  auto mangledName = swift_getTypeName(base, true);
+
+  auto getFieldAt = [&](const FieldDescriptor &descriptor) {
+    auto &field = descriptor.getFields()[index];
+    auto name = field.getFieldName(0);
+    auto type = field.getMangledTypeName(0);
+
+    auto description = base->getNominalTypeDescriptor();
+    if (!description)
+      return;
+
+    auto &genericParams = description->GenericParams;
+    auto numberOfLevels = genericParams.NestingDepth;
+    size_t paramsPerLevel[numberOfLevels];
+
+    for (unsigned i = 0; i < numberOfLevels; ++i) {
+      auto numParams = description->getGenericContext(i).NumPrimaryParams;
+      paramsPerLevel[i] = numParams;
+    }
+
+    auto *typeMetadata =
+        swift_getTypeByMangledName(type.data(), type.length(), numberOfLevels,
+                                   paramsPerLevel, base->getGenericArgs());
```

Comment:
@DougGregor Is this the correct way of using `getGenericArgs`? I've also modified interface a bit to use a callback instead of trying to return `std::pair`, WDYT?

Reply:
Since you're using a callback, you might also be able to make the callback use `StringRef` instead of `std::string`, since it could have access to memory from `swift_getFieldAt`'s stack frame.

Reply:
Done!